File: LeetCode Solutions/April 2021 LeetCoding Challenge/Palindrome Linked List/Palindrome Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode slow = head, fast = head, previous = null;
        while (fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        previous = slow;
        slow = slow.next;
        previous.next = null;
        while (slow != null){
            ListNode temp = slow.next;
            slow.next = previous;
            previous = slow;
            slow = temp;
        }
        fast = head;
        slow = previous;
        while (slow != null){
            if (fast.val != slow.val){
                return false;
            }
            fast = fast.next;
            slow = slow.next;
        }
        return true;
    }
}

File: LeetCode Solutions/April 2021 LeetCoding Challenge/Palindrome Linked List/Palindrome Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head;
        ListNode* previous = NULL;
        while (fast != NULL && fast->next != NULL){
            slow = slow->next;
            fast = fast->next->next;
        }
        previous = slow;
        slow = slow->next;
        previous->next = NULL;
        while (slow != NULL){
            ListNode* temp = slow->next;
            slow->next = previous;
            previous = slow;
            slow = temp;
        }
        fast = head;
        slow = previous;
        while (slow != NULL){
            if (fast->val != slow->val){
                return false;
            }
            fast = fast->next;
            slow = slow->next;
        }
        return true;
    }
};